# Homework 5

### Question 1

$$ \nabla f(x) = 
\begin{pmatrix}
\frac{\partial f}{\partial x_1} \\ \frac{\partial f}{\partial x_2}
\end{pmatrix} 
=\begin{pmatrix}
\ 2(200x_1^3 - 200x_1x_2 + x_1 - 1) \\ 200(x_2 - x_1^2)
\end{pmatrix} $$ 

$$ \nabla^2 f(x) = 
\begin{pmatrix}
\frac{\partial ^2 f}{\partial x_1} & \frac{\partial f}{\partial x_1x_2}\\
\frac{\partial f}{\partial x_2x_1} & \frac{\partial ^2 f}{\partial x_2}
\end{pmatrix} 
=\begin{pmatrix}
2(600x_1^2 - 200x_2 + 1) & -400x_1\\
-400x_1 & 200
\end{pmatrix}$$ 

### Question 2

In [1]:
import numpy as np

In [2]:
def f(x):
    x_1 = x[0][0]
    x_2 = x[1][0]
    return 100*(x_2 - x_1**2)**2 + (1-x_1)**2

def gradient(x):
    x_1 = x[0][0]
    x_2 = x[1][0]
    
    df_dx1 = 2*(200*x_1**3 - 200*x_1*x_2 + x_1 - 1)
    df_dx2 = 200*(x_2 - x_1**2)
    return np.array([[df_dx1],[df_dx2]])

def hessian(x):
    x_1 = x[0][0]
    x_2 = x[1][0]
    
    d2f_dx1 = 2*(600*x_1**2 - 200*x_2 + 1)
    df_dx1dx2 = -400*x_1
    d2f_dx2 = 200
    df_dx2dx1 = -400*x_1 
    return np.array([[d2f_dx1, df_dx1dx2],[df_dx2dx1, d2f_dx2]])

$$
x_{0} = \begin{pmatrix}
1.2 \\ 1.2
\end{pmatrix} 	
$$

Newton's Method converges in 7 steps.

In [3]:
# Initialize variables
x = np.array([[1.2],[1.2]])
k = 0
error = 10**-4
p = 0.9 # in (0,1)
c = 10**-5 # in (0,1)

gradient_norm = np.linalg.norm(gradient(x))
d = -np.linalg.inv(hessian(x))@gradient(x)

################################################################################

while gradient_norm > error:
    print("Iteration " + str(k))
    print("x_" + str(k) + " = " + str(np.matrix(x)))
    
    a = 1
    
    while f(x + a*d) > (f(x) + c*a*(gradient(x).transpose()@d)):
        a = p*a
   
    print("a_" + str(k) + " = " + str(a))
    print("")
    
    x = x + a*d    
    d = -np.linalg.inv(hessian(x))@gradient(x)
    gradient_norm = np.linalg.norm(gradient(x))
    k+=1

Iteration 0
x_0 = [[1.2]
 [1.2]]
a_0 = 1

Iteration 1
x_1 = [[1.19591837]
 [1.43020408]]
a_1 = 0.6561000000000001

Iteration 2
x_2 = [[1.0678032 ]
 [1.12378445]]
a_2 = 1

Iteration 3
x_3 = [[1.05197555]
 [1.10640204]]
a_3 = 1

Iteration 4
x_4 = [[1.00247988]
 [1.00251608]]
a_4 = 1

Iteration 5
x_5 = [[1.00081549]
 [1.00162887]]
a_5 = 1

Iteration 6
x_6 = [[1.00000045]
 [1.00000024]]
a_6 = 1



$$
x_{0} = \begin{pmatrix}
-1.2 \\ 1.0
\end{pmatrix} 	
$$

Newton's Method converges in 19 steps.

In [4]:
# Initialize variables
x = np.array([[-1.2],[1]])
k = 0
error = 10**-4
p = 0.9 # in (0,1)
c = 10**-5 # in (0,1)

gradient_norm = np.linalg.norm(gradient(x))
d = -np.linalg.inv(hessian(x))@gradient(x)

################################################################################

while gradient_norm > error:
    print("Iteration " + str(k))
    print("x_" + str(k) + " = " + str(np.matrix(x)))
    
    a = 1
    
    while f(x + a*d) > (f(x) + c*a*(gradient(x).transpose()@d)):
        a = p*a
   
    print("a_" + str(k) + " = " + str(a))
    print("")
    
    x = x + a*d    
    d = -np.linalg.inv(hessian(x))@gradient(x)
    gradient_norm = np.linalg.norm(gradient(x))
    k+=1

Iteration 0
x_0 = [[-1.2]
 [ 1. ]]
a_0 = 1

Iteration 1
x_1 = [[-1.1752809 ]
 [ 1.38067416]]
a_1 = 0.16677181699666577

Iteration 2
x_2 = [[-0.85201111]
 [ 0.62091045]]
a_2 = 1

Iteration 3
x_3 = [[-0.76783834]
 [ 0.58249067]]
a_3 = 0.38742048900000015

Iteration 4
x_4 = [[-0.48447315]
 [ 0.15007825]]
a_4 = 1

Iteration 5
x_5 = [[-0.40166751]
 [ 0.15448002]]
a_5 = 0.47829690000000014

Iteration 6
x_6 = [[-0.11895435]
 [-0.06935379]]
a_6 = 1

Iteration 7
x_7 = [[-0.0557394 ]
 [-0.00088925]]
a_7 = 0.43046721000000016

Iteration 8
x_8 = [[ 0.19684765]
 [-0.02732715]]
a_8 = 1

Iteration 9
x_9 = [[0.25334708]
 [0.06099256]]
a_9 = 0.47829690000000014

Iteration 10
x_10 = [[0.47131197]
 [0.1729609 ]]
a_10 = 1

Iteration 11
x_11 = [[0.52010727]
 [0.26813059]]
a_11 = 0.5904900000000002

Iteration 12
x_12 = [[0.71206808]
 [0.46921697]]
a_12 = 1

Iteration 13
x_13 = [[0.74568615]
 [0.55491766]]
a_13 = 0.7290000000000001

Iteration 14
x_14 = [[0.8969011 ]
 [0.78125934]]
a_14 = 1

Iteration 15
x_15

### Question 3

$$min \sum^3_{i=1}\sum^{5}_{j=1}\,c_{ij} x_{ij} 
= c_{AE}x_{AE} + c_{EH}x_{EH} + c_{AD} x_{AD} + c_{EF}x_{EF} + c_{HF}x_{HF} + c_{CH}x_{CH} + c_{CF}x_{CF} + c_{CG}x_{CG} + c_{FG}x_{FG} + c_{BG}x_{BG} + c_{BF}x_{BF} + c_{BD}x_{BD} + c_{FD}x_{FD}  \\
= 4x_{AE} + 2x_{EH} + 2x_{AD} + 4x_{EF} + 4x_{HF} + 2x_{CH} + 5x_{CF} + x_{CG} + x_{FG} + 2x_{BG} + 3x_{BF} + 3x_{BD} + 2x_{FD} $$


$$\sum^{3}_{i=1}\,x_{ij} \geq d_{j} \; \forall j\\
x_{AD} + x_{BD} + x_{CD} \geq 50 \\
x_{AE} + x_{BE} + x_{CE} \geq 60\\
x_{AF} + x_{BF} + x_{CF} \geq 40\\
x_{AG} + x_{BG} + x_{CG} \geq 30\\
x_{AH} + x_{BH} + x_{CH} \geq 70\\
$$	

$$\sum^{5}_{j=1}\,x_{ij} \leq s_{i} \; \forall i\\
x_{AD} + x_{AE} + x_{AF} + x_{AG} + x_{AH} \leq 100 \\
x_{BD} + x_{BE} + x_{BF} + x_{BG} + x_{BH} \leq 100 \\
x_{CD} + x_{CE} + x_{CF} + x_{CG} + x_{CH} \leq 80 \\
$$

$$x_{ij} \geq 0, i=1,2,3, j=1,2,3,4,5 $$


#### The final linear program is:  
$$min \; 4x_{AE} + 2x_{EH} + 2x_{AD} + 4x_{EF} + 4x_{HF} + 2x_{CH} + 5x_{CF} + x_{CG} + x_{FG} + 2x_{BG} + 3x_{BF} + 3x_{BD} + 2x_{DF} $$

such that $$x_{AD} + x_{BD} + x_{CD} \geq 50 \\
x_{AE} + x_{BE} + x_{CE} \geq 60\\
x_{AF} + x_{BF} + x_{CF} \geq 40\\
x_{AG} + x_{BG} + x_{CG} \geq 30\\
x_{AH} + x_{BH} + x_{CH} \geq 70\\
x_{AD} + x_{AE} + x_{AF} + x_{AG} + x_{AH} \leq 100 \\
x_{BD} + x_{BE} + x_{BF} + x_{BG} + x_{BH} \leq 100 \\
x_{CD} + x_{CE} + x_{CF} + x_{CG} + x_{CH} \leq 80 \\
\\
x_{ij} \geq 0, i=1,2,3, j=1,2,3,4,5 $$

**In terms of the pipe numbers:**

$$min \; 4x_{3} + 2x_{9} + 2x_{1} + 4x_{7} + 4x_{10} + 2x_{13} + 5x_{11} + x_{12} + x_{8} + 2x_{6} + 3x_{5} + 3x_{2} + 2x_{4} $$

i: let A = 1, B = 2, C = 3  
j: let D = 1, E = 2, F = 3, G = 4, H = 5

such that $$x_{1,1} + x_{2,1} + x_{3,1} \geq 50 \\
x_{1,2} + x_{2,2} + x_{3,2} \geq 60\\
x_{1,3} + x_{2,3} + x_{3,3} \geq 40\\
x_{1,4} + x_{2,4} + x_{3,4} \geq 30\\
x_{1,5} + x_{2,5} + x_{3,5} \geq 70\\
x_{1,1} + x_{1,2} + x_{1,3} + x_{1,4} + x_{1,5} \leq 100 \\
x_{2,1} + x_{2,2} + x_{2,3} + x_{2,4} + x_{2,5}  \leq 100 \\
x_{3,1} = x_{3,2} + x_{3,3} + x_{3,4} + x_{3,5}  \leq 80 \\
\\
x_{ij} \geq 0, i=1,2,3, j=1,2,3,4,5 $$

### Question 4

##### Part (a)
$$min \sum^{|G|}_{i=1}c_{i} p_{i} = 16p_1 + 20p_2 + 8p_3$$

such that $$\sum^{}_{j \in O(i)} f_{ij} - \sum^{}_{j \in I(i)} f_{ij}  = p_i \;\; for \; i \in G $$

$$p_1 = f_{12} -  f_{61} - 0 \\
p_2 = -f_{23} + f_{34} - 0 \\
p_3 = f_{56} - f_{45} - 0 $$
Note: I explicitly show subtraction of 0 to show that there's no inflow to the generator nodes. Also, some of the signs are negative because the outflow is flowing in the opposite direction of the notation.

$$\sum^{}_{j \in O(i)} f_{ij} - \sum^{}_{j \in I(i)} f_{ij} = -d_i\;\; for \; i \in D $$
$$d_1 = -110 = 0 - (f_{12} - f_{23}) \Rightarrow	-110 = -f_{12} + f_{23}\\
d_2 = -65 = 0 - (f_{34} - f_{45}) \Rightarrow -65 = -f_{34} + f_{45}\\
d_3 = -95 = 0 - (f_{56} - f_{61}) \Rightarrow -95 = -f_{56} + f_{61}$$
Note: Some of the signs are negative because the inflow is flowing in the opposite direction of the notation.

$$f_{ij} = B_{ij}(\theta_{i} - \theta_{j}) \; for \; all \; (i,j) \in E$$
$$ f_{12} = B_{12}(\theta_{1} - \theta_{2}) = 11.6(\theta_{1} - \theta_{2})\\
f_{23} = B_{23}(\theta_{2} - \theta_{3}) = 5.9(\theta_{2} - \theta_{3}) \\
f_{34} = B_{34}(\theta_{3} - \theta_{4}) = 13.7(\theta_{3} - \theta_{4})\\
f_{45} = B_{45}(\theta_{4} - \theta_{5}) = 9.8(\theta_{4} - \theta_{5}) \\
f_{56} = B_{56}(\theta_{5} - \theta_{6}) = 5.6(\theta_{5} - \theta_{6})\\
f_{61} = B_{61}(\theta_{6} - \theta_{1}) = 10.5(\theta_{6} - \theta_{1})
$$

$$-F_{ij} \leq f_{ij} \le F_{ij} \; for \; all \; (i,j) \in E$$
$$ -100 \leq f_{12} \leq 100 \\ 
-110 \leq f_{23} \leq 110 \\ 
-50 \leq f_{34} \leq 50 \\ 
-80 \leq f_{45} \leq 80 \\ 
-60 \leq f_{56} \leq 60 \\ 
-40 \leq f_{61} \leq 40 \\ 
$$


$$p^{m}_{i} \leq p_i \le p^{M}_{i} \; for \; all \; i \in G$$
$$20 \leq p_1 \leq 200 \\
20 \leq p_2 \leq 150 \\
10 \leq p_3 \leq 150$$

#### The final linear program is: 
$$min \; 16p_1 + 20p_2 + 8p_3$$
such that $$p_1 = f_{12} -  f_{61}, p_2 = f_{34} - f_{23}, p_3 = f_{56} - f_{45}$$

$$f_{23} - f_{12} = -110, f_{45} - f_{34} = -65, f_{61} - f_{56} = -95$$

$$f_{12} = 11.6(\theta_{1} - \theta_{2})\\
f_{23} = 5.9(\theta_{2} - \theta_{3}) \\
f_{34} = 13.7(\theta_{3} - \theta_{4})\\
f_{45} = 9.8(\theta_{4} - \theta_{5}) \\
f_{56} = 5.6(\theta_{5} - \theta_{6})\\
f_{61} = 10.5(\theta_{6} - \theta_{1})$$

$$-100 \leq f_{12} \leq 100 \\ 
-110 \leq f_{23} \leq 110 \\ 
-50 \leq f_{34} \leq 50 \\ 
-80 \leq f_{45} \leq 80 \\ 
-60 \leq f_{56} \leq 60 \\ 
-40 \leq f_{61} \leq 40 $$ 

$$20 \leq p_1 \leq 200 \\
20 \leq p_2 \leq 150 \\
10 \leq p_3 \leq 150
$$


##### Part (b)

In [5]:
import cvxpy as cp

# Define Objective Function
c = np.array([15,20,8])
p = cp.Variable(shape=(3,1), name="p")
obj_function = cp.matmul(c, p)
objective = cp.Minimize(cp.sum(obj_function, axis=0))

# p contraints
p_lower = np.array([[20],[20],[10]])
p_upper = np.array([[200],[150],[150]])

f = cp.Variable(shape=(1,6), name="f")
theta = cp.Variable(shape=(1,6), name="theta")

# f constraints
f_lower = np.array([[-100],[-110],[-50],[-80],[-60],[-40]])
f_upper = np.array([[100],[110],[50],[80],[60],[40]])

# formulas for [p1,p2,p3]
supply = f@np.array([[1,0,0],[0,-1,0],[0,1,0],[0,0,-1],[0,0,1],[-1,0,0]])

# formulas for [d1,d2,d3]
d = np.array([[-110],[-65],[-95]])
demand = f@np.array([[-1,0,0],[1,0,0],[0,-1,0],[0,1,0],[0,0,-1],[0,0,1]])

theta_formulas = theta@np.array([[11.6,0,0,0,0,-10.5],
          [-11.6,5.9,0,0,0,0],
          [0,-5.9,13.7,0,0,0],
          [0,0,-13.7,9.8,0,0],
          [0,0,0,-9.8,5.6,0],
          [0,0,0,0,-5.6,10.5]])

constraints = [ p >= p_lower.transpose(), 
                p <= p_upper.transpose(), 
                f >= f_lower.transpose(), 
                f <= f_upper.transpose(), 
                p == supply.T,
                d == demand.T,
                f.T == theta_formulas.T
              ]

problem = cp.Problem(objective, constraints)

In [6]:
problem.solve()

3191.840097245871

In [7]:
p.value

array([[113.12001378],
       [ 20.00000006],
       [136.87998615]])

In [8]:
f.value

array([[ 78.12001376, -31.87998624, -11.87998617, -76.87998617,
         59.99999998, -35.00000002]])

In [9]:
theta.value

array([[ 0.92212606, -5.81235789, -0.40897039,  0.45818189,  8.30307843,
        -2.41120728]])

The minimum cost to this electric power network is $3191.84.

$$p_1 = 113.12, p_2 = 20, p_3 = 136.88$$
$$f_{12} = 78.12, f_{23} = -31.88, f_{34} = -11.88, f_{45} = -76.88, f_{56} = 60, f_{61} = -35$$
$$\theta_{1} = 0.92, \theta_{2} = -5.81, \theta_{3} = -0.41, \theta_{4} = 0.46, \theta_{5} = 8.3, \theta_{5} = -2.41 $$

##### Part (c)

In [10]:
constraints[5].dual_value

array([[13.59901667],
       [ 9.65830686],
       [16.54775308]])

Elasticity prices for demand are as follows:
$$Node \;2 : \pi_1 = $13.60 \\
Node \;4 : \pi_2 = $9.66 \\
Node \;6 : \pi_3 = $16.54 $$